client/components/AdminPanel.jsx

```diff
@@ -0,0 +1,165 @@
+        super(props);
+
+        this.state = {
+            showModal: false,
+            isCreatingNewUser: false,
+            selectedUserToUpdate: {},
+            filterString: ''
+        };
+
+        this.fetchPage = this.fetchPage.bind(this);
+        this.openUpdateModal = this.openUpdateModal.bind(this);
+        this.onSubmitModal = this.onSubmitModal.bind(this);
+        this.submitCreateNewUser = this.submitCreateNewUser.bind(this);
+        this.submitUpdateUser = this.submitUpdateUser.bind(this);
+        this.closeModal = this.closeModal.bind(this);
+        this.deleteUser = this.deleteUser.bind(this);
+    }
+
+    componentDidMount() {
+        // Fetch the first page of users when the component mounts
+        this.props.fetchUsers(1, null);
+    }
+
+    fetchPage(page) {
+        const filter = this.state.filterString !== '' ? this.state.filterString : null;
+        this.props.fetchUsers(page, filter);
+    }
+
+    openUpdateModal(user) {
+        this.setState({
+            showModal: true,
+            selectedUserToUpdate: user
+        });
+    }
+
+    onSubmitModal(user) {
+        if (this.state.isCreatingNewUser) {
+            this.submitCreateNewUser(user);
+        } else {
+            this.submitUpdateUser(user);
+        }
+    }
+
+    async submitCreateNewUser(user) {
+        const { email, password, isAdmin } = user;
+        await this.props.adminCreateUser(email, password, isAdmin);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+
+        this.setState({
+            isCreatingNewUser: false
+        });
+
+        this.closeModal();
+    }
+
+    async submitUpdateUser(user) {
+        const { email, password, isAdmin } = user;
+        const userId = this.state.selectedUserToUpdate._id;
+        await this.props.updateUser(userId, email, password, isAdmin);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+
+        this.closeModal();
+    }
+
+    closeModal() {
+        this.setState({
+            showModal: false,
+            selectedUserToUpdate: {}
+        });
+    }
+
+    async deleteUser(user) {
+        await this.props.deleteUser(user._id);
+
+        // Refetch the users
+        this.fetchPage(this.props.currentPage);
+    }
+
+    render() {
+        if (!this.props.users) {
+            return <div></div>;
+        }
+
+        const { totalPages, totalEntries, currentPage, users } = this.props;
+
+        return (
+            <div>
+                <header className="navbar my-2">
+                    <section className="navbar-section">
+                        <div className="input-group input-inline">
+                            <input 
+                                className="form-input" type="text" placeholder="Filter users" 
+                                value={this.state.filterString}
+                                onChange={event => this.setState({ filterString: event.target.value })}
```

Comment:
this creates a new function for this handler every time this component renders

what's the better way to do this? hint, 2 sets of fat arrows.